### PR TITLE
Fix crash caused by reverse video playback and VideoSynchronizer tests

### LINKED ISSUE #6765: Video materials in reverse don't behave as expected
Video materials synchronized to the clock and then played in reverse don't behave as expected (reverse playback).

In Edge this seems to manifest as segments of the video running forward but ordered backwards.
In Chrome and Firefox this manifests for me as a frozen context.

https://cesiumjs.org/Cesium/Build/Apps/Sandcastle/?src=Video.html

### PR DESCRIPTION
**Pull request by lukesanantonio**
To follow up on #6765, I muted the sandcastle video and handled the `NotSupportedError` that's thrown when setting `playbackRate` to a bad value (this is what was causing the context freezes in both Firefox and Chrome).

I added a sync tolerance of 0.1 to the sandcastle so that seeking backwards isn't insanely choppy. I would have made it smaller but this seemed like a good middle ground value for all the browsers that I tested (Safari, Firefox, and Chrome). Chrome was overall very smooth for both normal and reverse playback.

As for Safari and Firefox, things get a little funky 🎸. When clock synchronization is enabled and there is a low tolerance set on the video sync, Safari and Firefox are both very choppy. At least during reverse playback, the choppiness is of the same degree. On the other hand, when a large tolerance is used, normal playback is smooth in both browsers, but reverse playback behaves completely differently! Safari is smooth because it handles negative playbackRate just fine and there is minimal hand-holding. Since Firefox doesn't support negative playbackRates, it becomes _very_ choppy because the `VideoSynchronizer` reverts to manual seeking. The minimal hand-holding has the opposite effect because the video is paused unlike Safari. I don't really like how the same tolerance value can mean very different things for each browser, but we have to handle the `NotSupportedError` somehow in order to stop the cesium viewer from crashing.

TL;DR

🎉 The context no longer freezes at all, so this should fix #6765! (Tested in Firefox, Chrome, and Safari). 😕 I don't know what's going on in Edge, but maybe lowering the tolerance would help.

🎉 I re-enabled the failing tests from #6772 and they _seem_ to be passing both locally and on Travis. 😕 I might be missing something here, but we'll see!

🤔 Another possibility is to add an optional `reverseTolerance` to `VideoSynchronizer` that applies only when syncing in reverse. It can be a bit tighter by default and would improve the choppiness in Firefox.

### CODE REVIEW

**Comment by cesium-concierge** [comment]
Thanks for the pull request @lukesanantonio!
* :heavy_check_mark: Signed CLA found.
* :grey_question: [CHANGES.md](https://github.com/AnalyticalGraphicsInc/cesium/blob/fix-video-sync/CHANGES.md) was not updated.
    * If this change updates the public API in any way, please add a bullet point to `CHANGES.md`.

Reviewers, don't forget to make sure that:

- [ ] Cesium Viewer works.
- [ ] Works in 2D/CV.
- [ ] Works (or fails gracefully) in IE11.

---

__I am a bot who helps you make Cesium awesome! [Contributions to my configuration are welcome.](https://github.com/AnalyticalGraphicsInc/cesium/tree/master/.concierge)__

:earth_africa: :earth_americas: :earth_asia:

**Comment by hpinkos** [comment]
Thanks @lukesanantonio!  I marked this `next release` since #6772 is marked `next release`

@mramato can you review when you have a chance?

**Comment by mramato** [comment]
Thanks for tracking this down @lukesanantonio!  I want to take a better look at this on Monday and read up on recent video changes in browsers a little more, but I think your general approach is good (especially in regards to muting the video).

**Comment by hpinkos** [comment]
@mramato reminder to review this for next release

File: Apps/Sandcastle/gallery/Video.html [hunk 1]
```diff
@@ -33,7 +33,7 @@
 <div id="loadingOverlay"><h1>Loading...</h1></div>
 <div id="toolbar"></div>
 
-<video id="trailer" style="display: none;" autoplay loop crossorigin controls>
+<video id="trailer" style="display: none;" muted autoplay loop crossorigin controls>
```
Comment (mramato) [inline Apps/Sandcastle/gallery/Video.html]: On my machine, the video wouldn't display until I actually showed the overlay, so let's remove `display:none` here and default the overlay to being shown for the best compatibility.

File: Source/Core/VideoSynchronizer.js [hunk 2]
```diff
@@ -172,7 +172,12 @@ define([
             return;
         }
 
-        element.playbackRate = clock.multiplier;
+        try {
+            element.playbackRate = clock.multiplier;
```
Comment (mramato) [inline Source/Core/VideoSynchronizer.js]: It's bad to have try/catch blocks in a hot area of the code, the optimizer can't do a good job and possibly throwing an exception every frame is really slow.  Is this only for negative rates?  Can we check once up front for negative values support and then just use a simple boolean here?

File: Apps/Sandcastle/gallery/Video.html [hunk 3]
```diff
@@ -75,7 +75,8 @@
 
     synchronizer = new Cesium.VideoSynchronizer({
         clock : viewer.clock,
-        element : videoElement
+        element : videoElement,
+        tolerance : 0.1
```
Comment (mramato) [inline Apps/Sandcastle/gallery/Video.html]: Let's remove this from the example, we should provide good defaults instead of immediately overriding things in the demo.

As for the defaults, I think for now it's OK to have a single tolerance value, and for the video to be choppy in reverse (reverse overall is an edge case anyway).  The problem with having a really low tolerance value is that seeking constantly starts to eat up CPU.  If we ever have a good use case for it, we can add a reverseTolerance in the future.

**Comment by mramato** [comment]
Thanks again @lukesanantonio!  Just those few comments.

File: Apps/Sandcastle/gallery/Video.html (hunk 1 above)
Comment (lukesanantonio) [reply Apps/Sandcastle/gallery/Video.html]: That makes sense, I'll set that up

File: Source/Core/VideoSynchronizer.js (hunk 2 above)
Comment (lukesanantonio) [reply Source/Core/VideoSynchronizer.js]: Yeah, good point. It's annoying but the exception can occur for large positive values too. Different browsers have different limits*. If we try to assign `playbackRate` out of the acceptable range without handling the exception, it seems to bubble up and cause the Viewer widget to freeze like in #6765.

What if we mark the value as invalid if an exception is thrown and then refuse to try again until it's changed? We can reset the actual `playbackRate` to 1.0 and let the clock adjustment take over. If in the `CesiumViewer` widget we ignore this exception specifically we can be a little less picky about what we assign to `playbackRate`.

\* (Chrome and Firefox only accept rates from 0.0625 to 16.0 according to [this SO answer](https://stackoverflow.com/a/32320020). I don't know about Safari or Edge, [but apparently playbackRate doesn't work at all on iOS](https://developer.apple.com/documentation/webkitjs/htmlmediaelement/1629746-playbackrate), we should probably keep that in mind.)

**Comment by lukesanantonio** [comment]
Thanks, @mramato! I fixed the smaller things you mentioned but still not sure what we should do with the try block.

File: Source/Core/VideoSynchronizer.js (hunk 2 above)
Comment (mramato) [reply Source/Core/VideoSynchronizer.js]: Caching the last value and only setting on change sounds like a good idea.  I would create a private helper function, `trySetPlaybackRate` that does it, so that the try/catch block is only inside that local helper and is a no-op if the old value and new value are the same.  You need to keep a separate `lastPlaybackRate` variable since there's no garuantee `element.playbackRate` will match the desired balue.

**Comment by lukesanantonio** [comment]
@mramato Check out this new change. Do you think there is room for testing some of this new behavior of using the last valid rate? What would make it difficult is that the limits are very browser specific.

**Comment by mramato** [comment]
> Do you think there is room for testing some of this new behavior of using the last valid rate?

I thought we were just setting playBackRate to 0 and using manually seeking, how come you changed it?

**Comment by lukesanantonio** [comment]
Oh that's a good point! Sorry, I think I just misinterpreted your comment. I just changed it back to what it was.

**Comment by mramato** [comment]
Looks good, I'll merge once this is green.  Thanks again @lukesanantonio!
